samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -29,97 +28,116 @@
    * transformed {@link MessageStream}.
    *
-   * @param mapFn the function to transform a {@link MessageEnvelope} to another {@link MessageEnvelope}
-   * @param <TM> the type of {@link MessageEnvelope}s in the transformed {@link MessageStream}
+   * @param mapFn the function to transform a message to another message
+   * @param <TM> the type of messages in the transformed {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM extends MessageEnvelope> MessageStream<TM> map(MapFunction<M, TM> mapFn);
+  <TM> MessageStream<TM> map(MapFunction<M, TM> mapFn);
 
   /**
-   * Applies the provided 1:n {@link FlatMapFunction} to transform a {@link MessageEnvelope} in this {@link MessageStream}
-   * to n {@link MessageEnvelope}s in the transformed {@link MessageStream}
+   * Applies the provided 1:n function to transform a message in this {@link MessageStream}
+   * to n messages in the transformed {@link MessageStream}
    *
-   * @param flatMapFn the function to transform a {@link MessageEnvelope} to zero or more {@link MessageEnvelope}s
-   * @param <TM> the type of {@link MessageEnvelope}s in the transformed {@link MessageStream}
+   * @param flatMapFn the function to transform a message to zero or more messages
+   * @param <TM> the type of messages in the transformed {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM extends MessageEnvelope> MessageStream<TM> flatMap(FlatMapFunction<M, TM> flatMapFn);
+  <TM> MessageStream<TM> flatMap(FlatMapFunction<M, TM> flatMapFn);
 
   /**
-   * Applies the provided {@link FilterFunction} to {@link MessageEnvelope}s in this {@link MessageStream} and returns the
+   * Applies the provided function to messages in this {@link MessageStream} and returns the
    * transformed {@link MessageStream}.
    * <p>
-   * The {@link FilterFunction} is a predicate which determines whether a {@link MessageEnvelope} in this {@link MessageStream}
+   * The {@link Function} is a predicate which determines whether a message in this {@link MessageStream}
    * should be retained in the transformed {@link MessageStream}.
    *
-   * @param filterFn the predicate to filter {@link MessageEnvelope}s from this {@link MessageStream}
+   * @param filterFn the predicate to filter messages from this {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
   MessageStream<M> filter(FilterFunction<M> filterFn);
 
   /**
-   * Allows sending {@link MessageEnvelope}s in this {@link MessageStream} to an output
-   * {@link org.apache.samza.system.SystemStream} using the provided {@link SinkFunction}.
+   * Allows sending messages in this {@link MessageStream} to an output using the provided {@link SinkFunction}.
    *
-   * @param sinkFn the function to send {@link MessageEnvelope}s in this stream to output systems
+   * NOTE: the output may not be a {@link org.apache.samza.system.SystemStream}. It can be an external database, etc.
```

Comment:
Keeping sink and sendTo separate makes sense, but I'm not so sure about the idea of terminal vs. not terminal (and thus, sendThrough). This distinction raises the question: is it possible we'd want to consume from the stream output to something not a stream but then pass the results through to something else. We could introduce a 4th function for this case, but I think it is easier to say that the distinction between terminal and not is clear in the way the graph is written and is otherwise unimportant. In other words, I would suggest making both sink and sendTo have the same type as the original stream and that they act like a tee operator.